rand_distr/src/zipf.rs

```diff
@@ -0,0 +1,366 @@
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> F {
+        let one = F::one();
+        loop {
+            let inv_b = self.inv_cdf(rng.sample(Standard));
+            let x = (inv_b + one).floor();
+            let mut ratio = x.powf(-self.s);
+            if x > one {
+                ratio = ratio * inv_b.powf(self.s)
+            };
+
+            let y = rng.sample(Standard);
+            if y < ratio {
+                return x;
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn test_samples<F: Float + core::fmt::Debug, D: Distribution<F>>(
+        distr: D, zero: F, expected: &[F],
+    ) {
+        let mut rng = crate::test::rng(213);
+        let mut buf = [zero; 4];
+        for x in &mut buf {
+            *x = rng.sample(&distr);
+        }
+        assert_eq!(buf, expected);
+    }
+
+    #[test]
+    #[should_panic]
+    fn zeta_invalid() {
+        Zeta::new(1.).unwrap();
+    }
+
+    #[test]
+    #[should_panic]
+    fn zeta_nan() {
+        Zeta::new(core::f64::NAN).unwrap();
+    }
+
+    #[test]
+    fn zeta_sample() {
+        let a = 2.0;
+        let d = Zeta::new(a).unwrap();
+        let mut rng = crate::test::rng(1);
+        for _ in 0..1000 {
+            let r = d.sample(&mut rng);
+            assert!(r >= 1.);
+        }
+    }
+
+    #[test]
+    fn zeta_small_a() {
+        let a = 1. + 1e-15;
+        let d = Zeta::new(a).unwrap();
+        let mut rng = crate::test::rng(2);
+        for _ in 0..1000 {
+            let r = d.sample(&mut rng);
+            assert!(r >= 1.);
+        }
+    }
+
+    #[test]
+    fn zeta_value_stability() {
+        test_samples(Zeta::new(1.5).unwrap(), 0f32, &[
+            1.0, 2.0, 1.0, 1.0,
+        ]);
+        test_samples(Zeta::new(2.0).unwrap(), 0f64, &[
+            2.0, 1.0, 1.0, 1.0,
+        ]);
+    }
+
+    #[test]
+    #[should_panic]
+    fn zipf_s_too_small() {
+        Zipf::new(10, -1.).unwrap();
+    }
+
+    #[test]
+    #[should_panic]
+    fn zipf_n_too_small() {
+        Zipf::new(0, 1.).unwrap();
+    }
+
+    #[test]
+    #[should_panic]
+    fn zipf_nan() {
+        Zipf::new(10, core::f64::NAN).unwrap();
+    }
+
+    #[test]
+    fn zipf_sample() {
+        let d = Zipf::new(10, 0.5).unwrap();
+        let mut rng = crate::test::rng(2);
+        for _ in 0..1000 {
+            let r = d.sample(&mut rng);
+            assert!(r >= 1.);
+        }
+    }
+
+    #[test]
+    fn zipf_sample_s_1() {
+        let d = Zipf::new(10, 1.).unwrap();
+        let mut rng = crate::test::rng(2);
+        for _ in 0..1000 {
+            let r = d.sample(&mut rng);
+            assert!(r >= 1.);
+        }
+    }
+
+    #[test]
+    fn zipf_sample_s_0() {
+        let d = Zipf::new(10, 0.).unwrap();
+        let mut rng = crate::test::rng(2);
+        for _ in 0..1000 {
+            let r = d.sample(&mut rng);
+            assert!(r >= 1.);
+        }
+        // TODO: verify that this is a uniform distribution
```

Comment:
Since we are almost finished with everything else, how would you like to verify this?

Should we verify the distribution as done in  [#1121](https://github.com/rust-random/rand/pull/1121)?
In this case, we know the pdf so it should be straightforward(?). 
I can help if you want.

Reply:
Sure, the approach from #1121 is the idea. Evaluating the zeta function might be a bit tricky, but I guess we can pull in a dev-dependency.

Maybe we can do this as a follow-up PR? We don't have this for most distributions at the moment, and this is easy to add later.

Reply:
Okay! no worries, I think this PR is done then :) (at least from my side)